### PR TITLE
[OCPSTRAT-506] Add CI for Azure Workload Identity disconnected.

### PR DESCRIPTION
**Pull request by huangmingxia**
Azure Managed Identity disconnected CI Support.
[CCO-380](https://issues.redhat.com/browse/CCO-380)
[CCO-409](https://issues.redhat.com/browse/CCO-409)

### CODE REVIEW

**Comment by huangmingxia** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.15-amd64-nightly-azure-ipi-workload-identity-tp-f14

File: ci-operator/step-registry/ipi/conf/azure/oidc-creds-provision/ipi-conf-azure-oidc-creds-provision-commands.sh [hunk 1]
```diff
@@ -146,6 +146,7 @@ ccoctl azure create-all \
   --region="${REGION}" \
   --subscription-id="${AZURE_SUBSCRIPTION_ID}" \
   --tenant-id="${AZURE_TENANT_ID}" \
+  --network-resource-group-name="${SHARED_DIR}/resourcegroup" \
```
Comment (jianlinliu) [inline ci-operator/step-registry/ipi/conf/azure/oidc-creds-provision/ipi-conf-azure-oidc-creds-provision-commands.sh:149]: I think here is some problems:
1. not each workflow calling this step would have `${SHARED_DIR}/resourcegroup`
2. In `azure-provision-resourcegroup` step, `${SHARED_DIR}/resourcegroup` are created to save the resource group naming for vnet creation, while the `ccoctl` command would also create an empty resource group with the same name for cluster install.

In a word, there are several resource groups for different purpose in this install, one is for vnet stuff, one is for cluster, another is for oidc storage.

Comment (jianlinliu) [reply ci-operator/step-registry/ipi/conf/azure/oidc-creds-provision/ipi-conf-azure-oidc-creds-provision-commands.sh:149]: And another thing is I think we should remove the line 131, installer should be able to create the metadata.json correctly.

Comment (huangmingxia) [reply ci-operator/step-registry/ipi/conf/azure/oidc-creds-provision/ipi-conf-azure-oidc-creds-provision-commands.sh:149]: Thank you. The PR [OCPBUGS-18246: Add networkResourceGroupName parameter for Azure](https://github.com/openshift/cloud-credential-operator/pull/597) for this feature has just been merged. I will debug it first. There is currently no nightly build available.

**Comment by huangmingxia** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.15-amd64-nightly-azure-ipi-disconnected-workload-identity-f14

**Comment by huangmingxia** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.15-amd64-nightly-azure-ipi-disconnected-workload-identity-f14

**Comment by huangmingxia** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.15-amd64-nightly-azure-ipi-disconnected-workload-identity-f14

**Comment by huangmingxia** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.15-amd64-nightly-azure-ipi-disconnected-workload-identity-f14

**Comment by huangmingxia** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.15-amd64-nightly-azure-ipi-disconnected-workload-identity-f14

File: ci-operator/step-registry/ipi/conf/azure/oidc-creds-provision/ipi-conf-azure-oidc-creds-provision-commands.sh [hunk 2]
```diff
@@ -139,6 +141,11 @@ if [ "${ENABLE_TECH_PREVIEW_CREDENTIALS_REQUESTS:-\"false\"}" == "true" ]; then
   ADDITIONAL_CCOCTL_ARGS="$ADDITIONAL_CCOCTL_ARGS --enable-tech-preview"
 fi
 
+if [ "${RESTRICTED_NETWORK:-\"no\"}" == "yes" ]; then
```
[the remark this answers is not in the document]
Comment (huangmingxia) [reply ci-operator/step-registry/ipi/conf/azure/oidc-creds-provision/ipi-conf-azure-oidc-creds-provision-commands.sh:144]: Thank you very much help to review, also learned a kind of new method, thank you very much. I appreciate it.

**Comment by huangmingxia** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.15-multi-nightly-azure-ipi-disconnected-workload-identity-f14

**Comment by huangmingxia** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.15-amd64-nightly-azure-ipi-disconnected-workload-identity-f14

**Comment by huangmingxia** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.15-multi-nightly-azure-ipi-disconnected-workload-identity-f14

**Comment by huangmingxia** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.15-multi-nightly-azure-ipi-disconnected-workload-identity-f14

**Comment by huangmingxia** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.15-multi-nightly-azure-ipi-oidc-disconnected-arm-f14

**Comment by huangmingxia** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.15-multi-nightly-azure-ipi-oidc-disconnected-amd-f28-destructive

File: ci-operator/step-registry/cucushift/installer/rehearse/azure/ipi/disconnected/private/cco-manual-workload-identity/deprovision/OWNERS [hunk 3]
```diff
@@ -0,0 +1,16 @@
+approvers:
+- jianlinliu
+- Amoghrd
+- yunjiang29
+- mgahagan73
```
Comment (jianlinliu) [inline ci-operator/step-registry/cucushift/installer/rehearse/azure/ipi/disconnected/private/cco-manual-workload-identity/deprovision/OWNERS:5]: Here you can remove `Amoghrd` `mgahagan73`, and replace `MayXuQQ` with @jinyunma

Comment (huangmingxia) [reply ci-operator/step-registry/cucushift/installer/rehearse/azure/ipi/disconnected/private/cco-manual-workload-identity/deprovision/OWNERS:5]: got it, thanks.

File: ci-operator/step-registry/cucushift/installer/rehearse/azure/ipi/disconnected/private/cco-manual-workload-identity/cucushift-installer-rehearse-azure-ipi-disconnected-private-cco-manual-workload-identity-workflow.metadata.json [hunk 4]
```diff
@@ -0,0 +1,23 @@
+{
+	"path": "cucushift/installer/rehearse/azure/ipi/disconnected/private/cco-manual-workload-identity/cucushift-installer-rehearse-azure-ipi-disconnected-private-cco-manual-workload-identity-workflow.yaml",
```
Comment (jianlinliu) [inline ci-operator/step-registry/cucushift/installer/rehearse/azure/ipi/disconnected/private/cco-manual-workload-identity/cucushift-installer-rehearse-azure-ipi-disconnected-private-cco-manual-workload-identity-workflow.metadata.json:2]: I suggest to use `disconnected` + `fullyprivate` to replace `disconnected` + `private`

File: ci-operator/config/openshift/openshift-tests-private/openshift-openshift-tests-private-release-4.15__multi-nightly.yaml [hunk 5]
```diff
@@ -250,6 +250,40 @@ tests:
...
     - chain: openshift-e2e-test-qe-destructive
     workflow: cucushift-installer-rehearse-azure-ipi-cco-manual-workload-identity
+- as: azure-ipi-oidc-disconnected-arm-f14
+  cron: 26 0 10,25 * *
+  steps:
+    cluster_profile: azure-qe
+    dependencies:
+      OPENSHIFT_INSTALL_RELEASE_IMAGE_OVERRIDE: release:multi-latest
+    env:
+      ADDITIONAL_WORKER_VM_TYPE: Standard_D4ps_v5
```
Comment (jianlinliu) [inline ci-operator/config/openshift/openshift-tests-private/openshift-openshift-tests-private-release-4.15__multi-nightly.yaml:260]: Should not need this setting.

File: ci-operator/config/openshift/openshift-tests-private/openshift-openshift-tests-private-release-4.15__multi-nightly.yaml [hunk 6]
```diff
@@ -250,6 +250,40 @@ tests:
...
+      BASE_DOMAIN: qe.azure.devcluster.openshift.com
+      E2E_RUN_TAGS: '@heterogeneous and @azure-ipi and @disconnected and @network-ovnkubernetes
```
Comment (jianlinliu) [inline ci-operator/config/openshift/openshift-tests-private/openshift-openshift-tests-private-release-4.15__multi-nightly.yaml:262]:
```suggestion
      E2E_RUN_TAGS: '@azure-ipi and @disconnected and @network-ovnkubernetes
```

File: ci-operator/config/openshift/openshift-tests-private/openshift-openshift-tests-private-release-4.15__multi-nightly.yaml [hunk 7]
```diff
@@ -250,6 +250,40 @@ tests:
...
+        and not @fips'
+      EXTRACT_MANIFEST_INCLUDED: "true"
+      TAG_VERSION: '@4.15'
+      TEST_FILTERS: ~ChkUpgrade&;~NonPreRelease&;~Serial&;~Disruptive&;~ConnectedOnly&;~HyperShiftMGMT&;~MicroShiftOnly&
+    test:
+    - ref: ipi-install-heterogeneous
```
Comment (jianlinliu) [inline ci-operator/config/openshift/openshift-tests-private/openshift-openshift-tests-private-release-4.15__multi-nightly.yaml:268]: Should not need this line.

File: ci-operator/config/openshift/openshift-tests-private/openshift-openshift-tests-private-release-4.15__multi-nightly.yaml [hunk 8]
```diff
@@ -250,6 +250,40 @@ tests:
...
+    workflow: cucushift-installer-rehearse-azure-ipi-disconnected-private-cco-manual-workload-identity
+- as: azure-ipi-oidc-disconnected-amd-f28-destructive
+  cron: 49 20 6 * *
+  steps:
+    cluster_profile: azure-qe
+    dependencies:
+      OPENSHIFT_INSTALL_RELEASE_IMAGE_OVERRIDE: release:multi-latest
+    env:
+      BASE_DOMAIN: qe.azure.devcluster.openshift.com
+      E2E_RUN_TAGS: '@heterogeneous and @azure-ipi and @disconnected and @network-ovnkubernetes
```
Comment (jianlinliu) [inline ci-operator/config/openshift/openshift-tests-private/openshift-openshift-tests-private-release-4.15__multi-nightly.yaml:279]:
```suggestion
      E2E_RUN_TAGS: '@azure-ipi and @disconnected and @network-ovnkubernetes
```

File: ci-operator/config/openshift/openshift-tests-private/openshift-openshift-tests-private-release-4.15__multi-nightly.yaml [hunk 9]
```diff
@@ -250,6 +250,40 @@ tests:
...
+        and not @fips'
+      EXTRACT_MANIFEST_INCLUDED: "true"
+      TAG_VERSION: '@4.15'
+      TEST_FILTERS: ~ChkUpgrade&;~NonPreRelease&;~Serial&;~Disruptive&;~ConnectedOnly&;~HyperShiftMGMT&;~MicroShiftOnly&
+    test:
+    - ref: ipi-install-heterogeneous
+    - chain: openshift-e2e-test-qe
+    workflow: cucushift-installer-rehearse-azure-ipi-disconnected-private-cco-manual-workload-identity
+- as: azure-ipi-oidc-disconnected-amd-f28-destructive
```
Comment (jianlinliu) [inline ci-operator/config/openshift/openshift-tests-private/openshift-openshift-tests-private-release-4.15__multi-nightly.yaml:271]:
```suggestion
- as: azure-ipi-disc-oidc-amd-f28-destructive
```

File: ci-operator/config/openshift/openshift-tests-private/openshift-openshift-tests-private-release-4.15__multi-nightly.yaml [hunk 10]
```diff
@@ -250,6 +250,40 @@ tests:
     test:
     - chain: openshift-e2e-test-qe-destructive
     workflow: cucushift-installer-rehearse-azure-ipi-cco-manual-workload-identity
+- as: azure-ipi-oidc-disconnected-arm-f14
```
Comment (jianlinliu) [inline ci-operator/config/openshift/openshift-tests-private/openshift-openshift-tests-private-release-4.15__multi-nightly.yaml:253]:
```suggestion
- as: azure-ipi-disc-fullyprivate-oidc-arm-f14
```

File: ci-operator/config/openshift/openshift-tests-private/openshift-openshift-tests-private-release-4.15__multi-nightly.yaml [hunk 11]
```diff
@@ -250,6 +250,40 @@ tests:
...
+    dependencies:
+      OPENSHIFT_INSTALL_RELEASE_IMAGE_OVERRIDE: release:multi-latest
+    env:
+      BASE_DOMAIN: qe.azure.devcluster.openshift.com
+      E2E_RUN_TAGS: '@heterogeneous and @azure-ipi and @disconnected and @network-ovnkubernetes
+        and not @fips'
+      EXTRACT_MANIFEST_INCLUDED: "true"
+      TAG_VERSION: '@4.15'
+    test:
+    - ref: ipi-install-heterogeneous
```
Comment (jianlinliu) [inline ci-operator/config/openshift/openshift-tests-private/openshift-openshift-tests-private-release-4.15__multi-nightly.yaml:284]: Should not need this line.

**Comment by huangmingxia** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.15-multi-nightly-azure-ipi-disc-oidc-amd-f28-destructive

**Comment by huangmingxia** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.15-multi-nightly-azure-ipi-disc-oidc-arm-f14

File: ci-operator/step-registry/cucushift/installer/rehearse/azure/ipi/disconnected/private/cco-manual-workload-identity/cucushift-installer-rehearse-azure-ipi-disconnected-private-cco-manual-workload-identity-workflow.metadata.json (hunk 4 above)
Comment (jianlinliu) [reply ci-operator/step-registry/cucushift/installer/rehearse/azure/ipi/disconnected/private/cco-manual-workload-identity/cucushift-installer-rehearse-azure-ipi-disconnected-private-cco-manual-workload-identity-workflow.metadata.json]: @huangmingxia I think you should reference cucushift-installer-rehearse-vsphere-ipi-disconnected-provision-chain.yaml to opt in the oidc part to complete the install workflow.

File: ci-operator/step-registry/cucushift/installer/rehearse/azure/ipi/disconnected/fullyprivate/cco-manual-workload-identity/provision/cucushift-installer-rehearse-azure-ipi-disconnected-fullyprivate-cco-manual-workload-identity-provision-chain.metadata.json [hunk 12]
```diff
@@ -0,0 +1,23 @@
+{
+	"path": "cucushift/installer/rehearse/azure/ipi/disconnected/fullyprivate/cco-manual-workload-identity/provision/cucushift-installer-rehearse-azure-ipi-disconnected-fullyprivate-cco-manual-workload-identity-provision-chain.yaml",
```
Comment (jianlinliu) [inline ci-operator/step-registry/cucushift/installer/rehearse/azure/ipi/disconnected/fullyprivate/cco-manual-workload-identity/provision/cucushift-installer-rehearse-azure-ipi-disconnected-fullyprivate-cco-manual-workload-identity-provision-chain.metadata.json:2]:
```suggestion
	"path": "cucushift/installer/rehearse/azure/ipi/disconnected/cco-manual-workload-identity/provision/cucushift-installer-rehearse-azure-ipi-disconnected-fullyprivate-cco-manual-workload-identity-provision-chain.yaml",
```

File: ci-operator/step-registry/cucushift/installer/rehearse/azure/ipi/disconnected/fullyprivate/cco-manual-workload-identity/provision/OWNERS [hunk 13]
```diff
@@ -0,0 +1,16 @@
+approvers:
+- jianlinliu
+- Amoghrd
```
Comment (jianlinliu) [inline ci-operator/step-registry/cucushift/installer/rehearse/azure/ipi/disconnected/fullyprivate/cco-manual-workload-identity/provision/OWNERS:3]: Here you can remove Amoghrd mgahagan73, and replace MayXuQQ with @jinyunma

File: ci-operator/config/openshift/openshift-tests-private/openshift-openshift-tests-private-release-4.15__multi-nightly.yaml [hunk 14]
```diff
@@ -250,6 +250,37 @@ tests:
     test:
     - chain: openshift-e2e-test-qe-destructive
     workflow: cucushift-installer-rehearse-azure-ipi-cco-manual-workload-identity
+- as: azure-ipi-disc-oidc-arm-f14
```
Comment (jianlinliu) [inline ci-operator/config/openshift/openshift-tests-private/openshift-openshift-tests-private-release-4.15__multi-nightly.yaml:253]: Do we also need the testing for 4.14?

File: ci-operator/step-registry/cucushift/installer/rehearse/azure/ipi/disconnected/fullyprivate/cco-manual-workload-identity/deprovision/cucushift-installer-rehearse-azure-ipi-disconnected-fullyprivate-cco-manual-workload-identity-deprovision-chain.yaml [hunk 15]
```diff
@@ -0,0 +1,10 @@
+chain:
+  as: cucushift-installer-rehearse-azure-ipi-disconnected-fullyprivate-cco-manual-workload-identity-deprovision
+  steps:
+    - ref: gather-installer-boostrap-logs-in-bastion
```
Comment (jianlinliu) [inline ci-operator/step-registry/cucushift/installer/rehearse/azure/ipi/disconnected/fullyprivate/cco-manual-workload-identity/deprovision/cucushift-installer-rehearse-azure-ipi-disconnected-fullyprivate-cco-manual-workload-identity-deprovision-chain.yaml:4]: I guess we do not this line here.

File: ci-operator/step-registry/cucushift/installer/rehearse/azure/ipi/disconnected/fullyprivate/cco-manual-workload-identity/provision/OWNERS (hunk 13 above)
Comment (huangmingxia) [reply ci-operator/step-registry/cucushift/installer/rehearse/azure/ipi/disconnected/fullyprivate/cco-manual-workload-identity/provision/OWNERS:3]: Sorry, missed it

File: ci-operator/config/openshift/openshift-tests-private/openshift-openshift-tests-private-release-4.15__multi-nightly.yaml (hunk 14 above)
Comment (huangmingxia) [reply ci-operator/config/openshift/openshift-tests-private/openshift-openshift-tests-private-release-4.15__multi-nightly.yaml:253]: yes, will add.

**Comment by huangmingxia** [comment]
periodic-ci-openshift-openshift-tests-private-release-4.15-multi-nightly-azure-ipi-disc-oidc-arm-f14
```
INFO[2023-10-07T13:41:01Z] Running step azure-ipi-disc-oidc-arm-f14-ipi-install-install. 
INFO[2023-10-07T14:28:20Z] Step azure-ipi-disc-oidc-arm-f14-ipi-install-install succeeded after 47m19s. 
```

**Comment by huangmingxia** [comment]
periodic-ci-openshift-openshift-tests-private-release-4.15-multi-nightly-azure-ipi-disc-oidc-amd-f28-destructive
```
INFO[2023-10-07T13:40:18Z] Running step azure-ipi-disc-oidc-amd-f28-destructive-ipi-install-install. 
INFO[2023-10-07T14:38:02Z] Step azure-ipi-disc-oidc-amd-f28-destructive-ipi-install-install succeeded after 57m43s. 
```
periodic-ci-openshift-openshift-tests-private-release-4.15-multi-nightly-azure-ipi-disc-oidc-arm-f14
```
INFO[2023-10-07T13:41:01Z] Running step azure-ipi-disc-oidc-arm-f14-ipi-install-install. 
INFO[2023-10-07T14:28:20Z] Step azure-ipi-disc-oidc-arm-f14-ipi-install-install succeeded after 47m19s. 
```

File: ci-operator/step-registry/cucushift/installer/rehearse/azure/ipi/disconnected/fullyprivate/cco-manual-workload-identity/deprovision/cucushift-installer-rehearse-azure-ipi-disconnected-fullyprivate-cco-manual-workload-identity-deprovision-chain.yaml (hunk 15 above)
Comment (huangmingxia) [reply ci-operator/step-registry/cucushift/installer/rehearse/azure/ipi/disconnected/fullyprivate/cco-manual-workload-identity/deprovision/cucushift-installer-rehearse-azure-ipi-disconnected-fullyprivate-cco-manual-workload-identity-deprovision-chain.yaml]: Thank you for your review, updated.

**Comment by huangmingxia** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.14-multi-nightly-azure-ipi-disc-oidc-arm-f14

**Comment by huangmingxia** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.14-multi-nightly-azure-ipi-disc-oidc-amd-f28-destructive

File: ci-operator/config/openshift/openshift-tests-private/openshift-openshift-tests-private-release-4.14__multi-nightly.yaml [hunk 16]
```diff
@@ -250,6 +250,37 @@ tests:
     test:
     - chain: openshift-e2e-test-qe-destructive
     workflow: cucushift-installer-rehearse-azure-ipi-cco-manual-workload-identity
+- as: azure-ipi-disc-oidc-arm-f14
+  cron: 26 13 12,27 * *
+  steps:
+    cluster_profile: azure-qe
+    dependencies:
+      OPENSHIFT_INSTALL_RELEASE_IMAGE_OVERRIDE: release:multi-latest
+    env:
```
Comment (jianlinliu) [inline ci-operator/config/openshift/openshift-tests-private/openshift-openshift-tests-private-release-4.14__multi-nightly.yaml:259]: For arm arch, I think need the following setting together:
```
      COMPUTE_NODE_TYPE: Standard_D4ps_v5
      OCP_ARCH: arm64
```

Comment (huangmingxia) [reply ci-operator/config/openshift/openshift-tests-private/openshift-openshift-tests-private-release-4.14__multi-nightly.yaml:259]: thank you.

**Comment by huangmingxia** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.14-multi-nightly-azure-ipi-disc-oidc-arm-f14

**Comment by huangmingxia** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.14-multi-nightly-azure-ipi-disc-oidc-amd-f28-destructive

**Comment by jianlinliu** [comment]
cc @liangxia to review.

**Comment by huangmingxia** [comment]
4.14-multi-nightly-azure-ipi-disc-oidc-amd-f28-destructive
```
INFO[2023-10-08T08:13:04Z] Running step azure-ipi-disc-oidc-amd-f28-destructive-ipi-install-install. 
INFO[2023-10-08T08:59:20Z] Step azure-ipi-disc-oidc-amd-f28-destructive-ipi-install-install succeeded after 46m15s. 
```
4.14-multi-nightly-azure-ipi-disc-oidc-arm-f14
```
INFO[2023-10-08T08:02:31Z] Running step azure-ipi-disc-oidc-arm-f14-ipi-install-install. 
INFO[2023-10-08T08:47:17Z] Step azure-ipi-disc-oidc-arm-f14-ipi-install-install succeeded after 44m45s. 
```

**Pull request merged**